Always terminate Monitor thread

xtrojak opened the pull request:
Added try-finally block to `Application.annotate_spectra` to ensure `Monitor` thread is always joined, even on some unknown exception. Although we catch all sorts of exceptions, there might always be a new one - in such case, it would pop up on top without `join` being called, resulting in "infinite" tool runs.

Close #86.

hechth reviewed (commented):
Seems fine - just also please add the info to the changelog in the PR.

I assume this is quite tricky to test, but would you have some idea how one could go about this?

xtrojak commented:
> Seems fine - just also please add the info to the changelog in the PR.
> 
> I assume this is quite tricky to test, but would you have some idea how one could go about this?

I suppose we would have to change the top-level code a little bit. If the `Monitor` and `Annotator` were passed to `Application.annotate_spectra`, we could mock them in a test and (maybe?) observe if mocked `Monitor` is terminated when mocked `Annotator` raises an exception. This would, however, influence how the top-level is called even in Galaxy wrapper.

hechth commented:
This means the constructor would change and we would inject the Monitor? I think this is fair, if it the constructor it can even have a default Monitor created there, that is fine if the call in the Galaxy wrapper has to change if that means we can actually test it.

xtrojak commented:
> This means the constructor would change and we would inject the Monitor? I think this is fair, if it the constructor it can even have a default Monitor created there, that is fine if the call in the Galaxy wrapper has to change if that means we can actually test it.

I tried to implement this in [2d2d161](https://github.com/RECETOX/MSMetaEnhancer/pull/87/commits/2d2d161ad8051a03ade03134ae808f97825aa800), let me know what you think about it. As you suggested, there are default objects created if they are not passed, so Galaxy wrapper does not need to be changed. I created "fake" `Monitor` and `Annotator` classes which can be used to mimic the behaviour of annotation process while allowing us to control if an exception is raised and whether the `Monitor` thread was terminated.

hechth commented on MSMetaEnhancer/app.py:
```diff
@@ -59,31 +59,40 @@ async def annotate_spectra(self, converters, jobs=None, repeat=False):
-                for converter in converters.values():
-                    jobs += converter.get_conversion_functions()
-            jobs = convert_to_jobs(jobs)
+            if annotator is None:
+                annotator = Annotator()
```
I think the constructor call can actually be moved to the default value for the argument.

hechth commented on MSMetaEnhancer/app.py:
```diff
@@ -59,31 +59,40 @@ async def annotate_spectra(self, converters, jobs=None, repeat=False):
+            annotator.set_converters(converters)
 
-            logger.set_target_attributes(jobs, len(self.spectra.spectrums))
+            if monitor is None:
+                monitor = Monitor()
```
Same goes to the monitor.

hechth commented on MSMetaEnhancer/libs/Annotator.py:
```diff
@@ -9,10 +9,13 @@ class Annotator:
     """
-    def __init__(self, converters):
-        self.converters = converters
+    def __init__(self):
+        self.converters = None
```
Is an annotator without converters actually functional? If not, then the parameter should I think still be in the constructor.

hechth reviewed (commented):
I think the default constructor calls (without conversions) can be moved to the default parameter values and it should be considered whether an annotator without any conversions is a valid object or not - if not, then they can also be instantiated before the `annotate_spectra` function and maybe have to be passed explicitly.

xtrojak commented:
> I think the default constructor calls (without conversions) can be moved to the default parameter values and it should be considered whether an annotator without any conversions is a valid object or not - if not, then they can also be instantiated before the `annotate_spectra` function and maybe have to be passed explicitly.

Good points, thanks! Actually `converters` can be empty (and in the case of `Monitor` it actually makes sense), I just changed their default value from `None` to empty `dict`. I moved constructor calls (without conversions) to the default parameter values of `annotate_spectra` as suggested.
